### PR TITLE
fix: Parse style when loading WFS from compositions

### PR DESCRIPTION
**Pull request by raitisbe**
## Description

Parse style from composition when loading WFS layers

## Related issues or pull requests

#2420 

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Dependency updates
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)
- [ ] I am unsure (we'll look into it together)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No
- [ ] I am unsure (no worries, we'll find out)

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the [MIT License]
- [x] I have followed the [guidelines for contributing](https://github.com/hslayers/hslayers-ng/blob/master/CONTRIBUTING.md)
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/hslayers/hslayers-ng/blob/master/CODE_OF_CONDUCT.md)
- [x] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally)
- [ ] I'm lost; why do I have to check so many boxes? Please help!

### CODE REVIEW

**Comment by raitisbe** [comment]
This should in theory solve it, I was just not successfull in finding any composition with a QML style in the composition definition to test on.

**Comment by fzadrazil** [comment]
You can try this composition, it still fails to load the style for its layers https://hub.lesprojekt.cz/rest/workspaces/fzadrazil/maps/demo2/file

**Comment by raitisbe** [comment]
yes, because `style` attribute is not set anywhere in the composition json->layers items. And https://hub.lesprojekt.cz/geoserver/fzadrazil/wfs points to wfs service which is XML.

It's different when we are adding layers from layman directly, since there we know the style url.

**Comment by raitisbe** [comment]
@fzadrazil Could it be that these compositions containig wfs were created in qgis? Because we are setting style attribute on layer in composition json when saving.

**Comment by fzadrazil** [comment]
> @fzadrazil We need to set the style attribute for layer in composition in qgis. Because in HSL we are setting style attribute on layer in composition json when saving.

Yes, I am just thinking where to put the style url in the composition schema. There is a params.styles property, but that is, I think, meant only for WMS style name. And property "style" is for explicit definition of style, not link to sld/qml. We could propose something in the context of this [Map Composition PR](https://github.com/hslayers/map-compositions/pull/1), eg. change the data type of "style" property so it is not an object, but could be also a string (url)?

```
"style": {
  "$id": "#/properties/layers/items/properties/style",
  "type": ["object", "string"],
  "title": "Style definition of vector layer features or url to style file (QML, SLD)",
  ...
}
```

**Comment by raitisbe** [comment]
> 

Yes, definitely. It thought it already is a string, since I'm storing SLD directly there. Object type was meant for the old way of storing styles pre-SLD era.
I can also check if the content begins with `http: ` and in that case do another request to fetch the style from server.

**Comment by raitisbe** [comment]
@fzadrazil Yup, I just checked it already is string in laymans version of the schema at https://github.com/LayerManager/layman/blob/master/src/layman/map/schema.draft-07.json
image
Seems, that I have forgotten to update it in our wiki and components/compositions/schema.json
Too many places..

And yes, we can't use WMS params STYLE.

**Review by fzadrazil — COMMENTED** [review]
I have requested the change in QGIS plugin/Layman. Once it is implemented, I will test again in hslayers.

**Comment by raitisbe** [comment]
It should load the style contents from a file defined in style: attribute of layer in composition.

**Comment by jmacura** [comment]
FTR: This feature is unit-tested but not live-tested in any real-world scenario. The PR was merged nevertheless in an attempt to be "future-proof". Once the feature is implemented in the QGIS Layman plugin, the functionality shall be retested.
